lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,607 @@
+  defp to_param(bin) when is_binary(bin), do: bin
+  defp to_param(false), do: "false"
+  defp to_param(true), do: "true"
+  defp to_param(data), do: Phoenix.Param.to_param(data)
+
+  defp route_type(router, statics, test_path) do
+    cond do
+      static_path?(test_path, statics) -> :static
+      match_route?(router, test_path) -> :match
+      true -> :error
+    end
+  end
+
+  defp match_route?(router, test_path) when is_binary(test_path) do
+    split_path = for segment <- String.split(test_path, "/"), segment != "", do: segment
+    match_route?(router, split_path)
+  end
+
+  defp match_route?(router, split_path) when is_list(split_path) do
+    case router.__match_route__(split_path) do
+      {nil = _forward_plug, _} -> true
+      {forward_plug, _} -> match_forward_route?(router, forward_plug, split_path)
+      :error -> false
+    end
+  end
+
+  defp match_forward_route?(router, forward_router, split_path) do
+    if function_exported?(forward_router, :__routes__, 0) do
+      script_name = router.__forward__(forward_router)
+      match_route?(forward_router, split_path -- script_name)
+    else
+      true
+    end
+  end
+
+  defp build_route(route_ast, og_ast, env, endpoint_ctx, router) do
+    statics = Module.get_attribute(env.module, :statics, [])
+    {endpoint_ctx, router, statics} = Macro.expand({endpoint_ctx, router, statics}, env)
+
+    {type, test_path, path_ast, static_ast} =
+      rewrite_path(route_ast, endpoint_ctx, router, statics)
+
+    route = %__MODULE__{
+      router: router,
+      stacktrace: Macro.Env.stacktrace(env),
+      inspected_route: Macro.to_string(og_ast),
+      test_path: test_path
+    }
+
+    {route, type, endpoint_ctx, route_ast, path_ast, static_ast}
+  end
+
+  defp rewrite_path(route, endpoint, router, statics) do
+    {rewrite_route, test_path} =
+      case route do
+        route when is_binary(route) ->
+          test_path =
+            case String.split(route, "?") do
+              [^route] -> route
+              [prefix, _query] -> prefix
+              _ -> raise ArgumentError, "invalid query string for path #{route}"
+            end
+
+          {route, test_path}
+
+        {:<<>>, meta, segments} = route ->
+          {path_rewrite, query_rewrite} = verify_segment(segments, route, [])
+          test_path = Enum.map_join(path_rewrite, &if(is_binary(&1), do: &1, else: "1"))
+          {{:<<>>, meta, path_rewrite ++ query_rewrite}, test_path}
+      end
+
+    type = route_type(router, statics, test_path)
+
+    path_ast =
+      quote do
+        unquote(__MODULE__).unverified_path(unquote_splicing([endpoint, router, rewrite_route]))
+      end
+
+    static_ast =
+      quote do
+        unquote(__MODULE__).static_path(unquote_splicing([endpoint, rewrite_route]))
+      end
+
+    {type, test_path, path_ast, static_ast}
+  end
+
+  defp attr!(env, name) do
+    Module.get_attribute(env.module, name) || raise "expected @#{name} module attribute to be set"
```

Comment:
I think we need a better error message here.

1. Router should always be given
2. The issue is that endpoint is optional but, if the endpoint is missing, we should tell people to use `path(conn_or_socket, ...)`. So a good error message should be tailored to endpoint and say that they should either pass it on use or use `path/2`